File: content/snippets/cpp/design_patterns_template_method.cpp
#include <iostream>

class DataProcessor {
public:
    virtual ~DataProcessor() = default;
    
    void process() {
        readData();
        parseData();
        writeData();
    }
    
protected:
    virtual void readData() = 0;
    virtual void parseData() = 0;
    virtual void writeData() = 0;
};

class CSVProcessor : public DataProcessor {
protected:
    void readData() override {
        std::cout << "Reading CSV data\n";
    }
    
    void parseData() override {
        std::cout << "Parsing CSV\n";
    }
    
    void writeData() override {
        std::cout << "Writing processed CSV\n";
    }
};

class JSONProcessor : public DataProcessor {
protected:
    void readData() override {
        std::cout << "Reading JSON data\n";
    }
    
    void parseData() override {
        std::cout << "Parsing JSON\n";
    }
    
    void writeData() override {
        std::cout << "Writing processed JSON\n";
    }
};

int main() {
    CSVProcessor csv;
    csv.process();
    
    JSONProcessor json;
    json.process();
}

File: content/snippets/python/design_patterns_template_method.py
from abc import ABC, abstractmethod

class DataProcessor(ABC):
    def process(self) -> None:
        self.read_data()
        self.parse_data()
        self.write_data()
    
    @abstractmethod
    def read_data(self) -> None:
        pass
    
    @abstractmethod
    def parse_data(self) -> None:
        pass
    
    @abstractmethod
    def write_data(self) -> None:
        pass

class CSVProcessor(DataProcessor):
    def read_data(self) -> None:
        print("Reading CSV data")
    
    def parse_data(self) -> None:
        print("Parsing CSV")
    
    def write_data(self) -> None:
        print("Writing processed CSV")

class JSONProcessor(DataProcessor):
    def read_data(self) -> None:
        print("Reading JSON data")
    
    def parse_data(self) -> None:
        print("Parsing JSON")
    
    def write_data(self) -> None:
        print("Writing processed JSON")

csv = CSVProcessor()
csv.process()

json = JSONProcessor()
json.process()
